### DISCUSSION TITLE
npm publish finishes before publish is done

### DISCUSSION THREAD

**Discussion by mafintosh** [npm/feedback#68] [Registry]
When `npm publish` finishes there is a grace period where the publish is not available through `npm install` - I assume it's probably propating out through the CDN.

This makes the bug fixing flow for dependencies more tedious than it needs to be as you have to wait 5-10 minutes before running CI on dependents after fixing something to make sure it's actually available through the install so you don't debug the same bug twice.

This might always have been the case but I've been feeling this more and more and it seems to be a decent amount longer than it used to be in the past.

Maybe we could have a command to check that the publish has actually been distributed or some flag to publish?

**Comment by MylesBorins** [comment]
There are definitely ways that we can explore improving this flow not limited to:

* improving time from publish to deploy
* explore a poll based publish strategy that would not exit `npm publish` until the publish is successful
* An update to end points to serve an alternative error code (e.g. 420) if a package has been publish but isn't available yet.

Thoughts?

FWIW `npm view` will show details about a package. You can specifically run `npm view ${package.name} version` to get the latest version of a package, you can append a dist-tag to get the version for the tag... e.g. `npm view npm@next version`.

**Comment by mafintosh** [comment]
I think some poll based approach would be good. Could even be opt-in, as it's probably mostly a poweruser feature.

Does the network give any guarantees about atomicity in terms of updates? Ie. when a poll to check if the version I just published on my computer finishes, does that mean it'll be available for anyone else when they do a new npm install?

**Reply by ethomson** [reply]
I expect that when `npm view` returns and shows the package that it should be installable.  Obviously though we're talking about race conditions _already_ so there may be nasty edge cases here, but I feel comfortable committing to this outcome.

**Reply by mafintosh** [reply]
Awesome! So we could just poll for that in publish `npm publish --wait-until-installable`

**Reply by MylesBorins** [reply]
This definitely seems like a potential solution. The only thing I would want to figure out in addition would be if we want to make any registry changes / introduce a different command that might be a more robust solution to the problem.

**Comment by andreialecu** [comment]
Here's a simple shell script that should work in the mean time:

```sh
# poll-npm.sh

PACKAGE=$1
VERSION=$2
until [ $(npm view $PACKAGE --json | jq --arg version "$VERSION" -r '.versions[] | select (. == $version)') ];
do
  echo "Waiting for publish to complete"
  sleep 5s
done
```

To run: `timeout 5m ./poll-npm.sh react 99.0.0`

**Reply by ljharb** [reply]
note that if you publish with a tag (not "latest") this might not work, but if you use `npm show $PACKAGE versions --json` it should be doable.

**Reply by MylesBorins** [reply]
FWIW `npm view ${package.name}@${package.version} version` will return the version number if it is available and an empty string if it isn't.
